File: src/leetcode/java/greedy/JumpGameII_45.java
package leetcode.java.greedy;

public class JumpGameII_45 {
  public int jump(int[] nums) {
    if (nums.length == 1) return 0;

    int cnt = 0, pridx, idx = 0;
    while (true) {
      if (idx + nums[idx] >= nums.length - 1) {
        cnt++;
        break;
      }
      pridx = idx;
      idx = findNextPos(nums, pridx + 1, pridx + nums[pridx]);
      cnt++;
    }
    return cnt;
  }

  private static int findNextPos(int[] nums, int start, int end) {
    int idx = start, max = -1;

    for (int i = end; i >= start; i--) {
      if (i + nums[i] >= nums.length - 1) return i;
      if (i + nums[i] > max) {
        idx = i;
        max = i + nums[i];
      }
    }
    return idx;
  }
}

File: src/leetcode/python/greedy/JumpGameII_45.py
class JumpGameII_45(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums)
        if length == 1:
            return 0

        index = 0
        count = 0
        while True:
            if nums[index] + index >= length - 1:
                count += 1
                break
            pre_index = index
            i = pre_index + nums[pre_index]
            maxtmp = -1
            while i >= pre_index:
                if i + nums[i] > maxtmp:
                    maxtmp = i + nums[i]
                    index = i
                i -= 1
            count += 1
        return count
